### PR TITLE
#1431 Rails 6: Move sprockets/assets over to npm/webpacker!

### PR DESCRIPTION
**Pull request by murny**
#1431

Includes FontAwesome v5 upgrade

TODO: 
- [x] Rebase off latest
- [x] ES6 Javascript
- [x] Update docker/CI infrastructure
- [x] ESLINT (installed/configured, but not enabled in CI, project not linted, etc)
- [x] Manual test heavily, lots of changes here

### CODE REVIEW

File: app/controllers/admin/communities_controller.rb [hunk 1]
```diff
@@ -10,8 +10,8 @@ def index
         render template: 'communities/index'
       end
       format.json do
-        community_title_index = Community.solr_name_for(:title, role: :search)
-        collection_title_index = Collection.solr_name_for(:title, role: :search)
+        community_title_index = Community.solr_exporter_class.solr_name_for(:title, role: :search)
```
Comment (murny) [inline app/controllers/admin/communities_controller.rb]: This was a bug I found, looks like we missed a couple `.solr_exporter_class`

Comment (mbarnett) [reply app/controllers/admin/communities_controller.rb]: Aha. We must be missing test coverage for this JSON response?

File: app/views/layouts/application.html.erb [hunk 2]
```diff
@@ -40,11 +40,10 @@
 
     <%= csrf_meta_tags %>
 
-    <%# TODO: Investigate CSP when more documentation with doing js.erb/turbolinks etc. %>
-    <%#= csp_meta_tag %>
+    <%= csp_meta_tag %>
```
Comment (murny) [inline app/views/layouts/application.html.erb]: I reenabled CSP, but allowing inline css/javascript (which was what is causing all the problems). Something is better than nothing!

File: app/controllers/admin/communities_controller.rb (hunk 1 above)
Comment (murny) [reply app/controllers/admin/communities_controller.rb]: Yeah must have no test coverage. It 500s.

Comment (mbarnett) [reply app/controllers/admin/communities_controller.rb]: Do you want to throw a really simple one on this? Even if it just looks for a 200 OK it'll at least catch regressions.

**Comment by murny** [comment]
Thanks @mbarnett! I believe everything we need Henry to do is being tracked on this ticket: https://github.com/ualbertalib/jupiter/issues/1433

I know there's a nodejs playbook and i know discovery uses it (as we had to switch over to it over `therubyracer` for a JS runtime). I assume Jupiter was using this way before discovery. I believe we will probably need `yarn` installed! Other then that  nothing * Should * change 🤞. The same `assets:precompile` rake tasks will work with webpacker

**Comment by mbarnett** [comment]
Cool, thanks!

File: README.md [hunk 3]
```diff
@@ -6,6 +6,16 @@
...
 
+**Requirements:**
+
+- **Ruby** 2.6+
+- **PostgreSQL**
+- **Redis**
+- **Solr**
+- **ImageMagick**
+- **Node.js**
+- **Yarn**
```
Comment (pgwillia) [inline README.md]: The apt packages on Ubuntu 18.04 didn't work but this did https://www.digitalocean.com/community/tutorials/how-to-install-node-js-on-ubuntu-18-04

File: config/application.rb [hunk 4]
```diff
@@ -12,7 +12,7 @@
 require 'action_text/engine'
 require 'action_view/railtie'
 # require "action_cable/engine"
-require 'sprockets/railtie'
+# require 'sprockets/railtie'
```
Comment (pgwillia) [inline config/application.rb]: Why comment out?

File: package.json [hunk 5]
```diff
@@ -1,5 +1,34 @@
...
+    "dropzone": "^5.7.0",
+    "easy-autocomplete": "^1.3.5",
+    "expose-loader": "^0.7.5",
+    "jquery": "^3.4.1",
+    "jquery-multidownload": "^4.0.1",
+    "popper.js": "^1.16.0",
+    "regenerator-runtime": "^0.13.3",
+    "selectize": "^0.12.6",
+    "trix": "^1.2.2",
+    "turbolinks": "^5.2.0"
```
Comment (pgwillia) [inline package.json]: This is going to be so much better!

File: config/application.rb (hunk 4 above)
Comment (murny) [reply config/application.rb]: Shows what railties are available and which ones we have we have opted out from. 

Helps with upgrading.

Typically if you wanted everything you would say `require 'rails/all'.

But here we specifically saying which ones we want or not want:

```
require 'rails'
# Pick the frameworks you want:
require 'active_model/railtie'
require 'active_job/railtie'
require 'active_record/railtie'
require 'active_storage/engine'
require 'action_controller/railtie'
require 'action_mailer/railtie'
require 'action_mailbox/engine'
require 'action_text/engine'
require 'action_view/railtie'
# require "action_cable/engine"
require 'sprockets/railtie'
require 'rails/test_unit/railtie'
```

File: .travis.yml [hunk 6]
```diff
@@ -31,4 +41,4 @@ script:
   - bundle exec rails test
   - bundle exec rails test:system
   - bundle exec rubocop
-  - bundle exec scss-lint
```
[the remark this answers is not in the document]
Comment (murny) [reply .travis.yml]: Yep I did that 👍

File: README.md [hunk 7]
```diff
@@ -6,6 +6,16 @@
 
 ![Jupiter Architecture Diagram](docs/jupiter_architecture.png)
 
+**Requirements:**
+
+- **Ruby** 2.6+
```
[the remark this answers is not in the document]
Comment (murny) [reply README.md]: Oops I though Rails 6 required Ruby 2.6...but you are correct! It's Ruby 2.5 

Can update that 😓

File: .travis.yml (hunk 6 above)
[the remark this answers is not in the document]
Comment (pgwillia) [reply .travis.yml]: Didn't see a change to Gemfile.lock?

File: config/application.rb (hunk 4 above)
Comment (pgwillia) [reply config/application.rb]: Cool, cool.

File: README.md (hunk 3 above)
Comment (murny) [reply README.md]: I'd highly recommend people to use a version manager for ruby and node. 

There's a few out there. NVM is popular one for Nodejs. RBenv/RVM is popular for Ruby. There is also some that work for multiple dependencies like ASDF (which is what I use).

Helps you switch between projects (that may require different node/ruby versions) etc.

File: .travis.yml (hunk 6 above)
[the remark this answers is not in the document]
Comment (murny) [reply .travis.yml]: Haha thanks for the second pair of eyes...totally remember doing that but you are right! Seems to still be there 🤦‍♂.

Will update this too 👍

[the remark this answers is not in the document]
Comment (murny) [reply .travis.yml]: Updated this now 😄

File: README.md (hunk 7 above)
[the remark this answers is not in the document]
Comment (murny) [reply README.md]: This is now updated, thanks!

File: README.md (hunk 3 above)
Comment (pgwillia) [reply README.md]: I'll check ASDF out!

**Pull request merged**